openhcl/virt_mshv_vtl/src/processor/snp/mod.rs

```diff
@@ -930,40 +928,46 @@ impl UhProcessor<'_, SnpBacked> {
     }
 
     async fn run_vp_snp(&mut self, dev: &impl CpuIo) -> Result<(), VpHaltReason<UhRunVpError>> {
-        // TODO CVM GUEST VSM: split out the "next vtl" that VTL 2 will exit to from
-        // last_vtl/current_vtl
-        let last_vtl = self
+        // TODO CVM GUEST VSM: actually check if there is an interrupt waiting
+        // for VTL 1 and switch to it if there is
+        let next_vtl = self
             .cvm_guest_vsm
             .as_ref()
-            .map_or(GuestVtl::Vtl0, |gvsm_state| gvsm_state.current_vtl);
+            .map_or(GuestVtl::Vtl0, |gvsm_state| gvsm_state.exit_vtl);
 
-        let mut vmsa = self.runner.vmsa_mut(last_vtl);
+        let mut vmsa = self.runner.vmsa_mut(next_vtl);
         let last_interrupt_ctrl = vmsa.v_intr_cntrl();
 
         vmsa.v_intr_cntrl_mut().set_guest_busy(false);
 
-        // TODO CVM GUEST VSM actually check and run vtl 1
         self.unlock_tlb_lock(Vtl::Vtl2);
-        let tlb_halt = self.should_halt_for_tlb_unlock(GuestVtl::Vtl0);
+        let tlb_halt = self.should_halt_for_tlb_unlock(next_vtl);
 
-        self.runner.set_halted(
-            self.backing.lapics[GuestVtl::Vtl0].halted
-                || self.backing.lapics[GuestVtl::Vtl0].startup_suspend
-                || tlb_halt,
-        );
+        let halt = self.backing.lapics[next_vtl].halted
```

Comment:
Do we need to check if vtl 1 is halted when running vtl 0? Can we even end up in that scenario?

Reply:
We maybe could consider a sanity check but I don't think we can run into that scenario. If VTL 1 was the last VTL to run (and I think it must have been in order to cause the halt), the exit handling code should never decide to go back to VTL 0 until VTL 1 says it can. So I think that means that the exit_vtl variable will stay at VTL 1, and the exit code should keep using VTL 1.